### PR TITLE
Expose alias multinomial methods to ATen

### PR DESCRIPTION
**Pull request by vishwakftw**
This PR exposes the multinomialAliasSetup and multinomialAliasDraw methods.

cc: @neerajprad

### CODE REVIEW

**Comment by vishwakftw** [comment]
TODO:
- [ ] I'll have to add checks for `num_samples` being >= 0.
- [ ] Tests (any tests you have in mind, @neerajprad?)

**Comment by neerajprad** [comment]
This is looking great! 

> Tests (any tests you have in mind, @neerajprad?)

I think we can test the sampler in more detail when we use it within distributions. For the time being, what do you think about adapting the multinomial tests in `test_torch.py` and `test_cuda.py`?

**Comment by vishwakftw** [comment]
I inspected the code in THTensorRandom.cpp and THCTensorRandom.cu in TH and THC respectively. They only seem to work on probability tensors of dim 1. Is this okay?

**Comment by vishwakftw** [comment]
I inspected the implementations of these functions in THTensorRandom.cpp and THCTensorRandom.cu under TH and THC respectively, and it seems that these methods only work for probability tensors of dim 1. Is this desirable?

**Comment by neerajprad** [comment]
> I inspected the code in THTensorRandom.cpp and THCTensorRandom.cu in TH and THC respectively. They only seem to work on probability tensors of dim 1. Is this okay?

I think this imposes a restriction in that we can't generally swap out the standard binary search implementation for the alias sampler. It makes sense though because building the alias table itself takes O(Nk), where N are the number of distributions and k the number of categories. We probably don't want to encourage users to use this with N > 1 (specially when N > 1 due to expanded distribution instances where we can get by with a single alias table). The standard sampler on the other hand can just treat N as n_sample, and can support different distributions easily. I think this restriction is fine, we should just throw some meaningful exception, and make sure that we are considering this in the `.expand` method for the categorical distribution later.

**Comment by neerajprad** [comment]
@vishwakftw - Let me know if you need any help with the tests, I could add to this PR itself.

**Comment by vishwakftw** [comment]
@neerajprad thanks! I have some tests ready locally, will push them in the morning. Sorry about the delay.

**Comment by neerajprad** [comment]
No worries at all! I was just checking in, please take your time.

**Comment by vishwakftw** [comment]
@neerajprad I've added a simple test based on the multinomial tests. Let me know if those are fine, thanks.

**Comment by neerajprad** [comment]
We are getting a test failure:

```
Mar 19 12:12:23 FAIL: test_multinomial_alias (__main__.TestTorch)
Mar 19 12:12:23 ----------------------------------------------------------------------
Mar 19 12:12:23 RuntimeError: _th_multinomial_alias_draw is not implemented for type torch.LongTensor
Mar 19 12:12:23 
Mar 19 12:12:23 During handling of the above exception, another exception occurred:
Mar 19 12:12:23 
Mar 19 12:12:23 Traceback (most recent call last):
Mar 19 12:12:23   File "test_torch.py", line 3308, in test_multinomial_alias
Mar 19 12:12:23     self._test_multinomial_alias(self, lambda t: t)
Mar 19 12:12:23   File "test_torch.py", line 3304, in _test_multinomial_alias
Mar 19 12:12:23     torch.multinomial_alias_draw(prob_table, alias_table, n_samples)
Mar 19 12:12:23 AssertionError: "cannot sample <= 0 samples" does not match "_th_multinomial_alias_draw is not implemented for type torch.LongTensor"
Mar 19 12:12:23 
```

File: test/test_torch.py [hunk 1]
```diff
@@ -3280,6 +3280,33 @@ def make_prob_dist(shape, is_contiguous):
...
+        def get_probs(length, is_contiguous):
+            probs = torch.softmax(torch.randn(length), 0)
+            if not is_contiguous:
+                probs = torch.softmax(torch.randn(length, 2), 0)[:, 1]
+            assert not (is_contiguous ^ probs.is_contiguous()), "contiguity requirement not met"
+            return cast(probs)
+
+        for is_contiguous in [True, False]:
+            probs = get_probs(4, is_contiguous)
+            prob_table, alias_table = torch.multinomial_alias_setup(probs)
```
Comment (neerajprad) [inline test/test_torch.py]: What would happen if probs is n-dimensional (n > 1)? Would this throw an exception?

**Comment by vishwakftw** [comment]
I'll look into the test failures, thank you for letting me know.

Regarding the n-d case, I realized that the function do accept n-d inputs. The existence of a `fastGet1d` function in the code mislead me into thinking that the function works for 1d inputs only. I'll add test cases for the multi-dim case as well.

**Comment by neerajprad** [comment]
> Regarding the n-d case, I realized that the function does accept n-d inputs. The existence of a fastGet1d function in the code mislead me into thinking that the function works for 1d inputs only. I'll add test cases for the multi-dim case as well.

That's awesome! That would really help keep things simple.

**Comment by vishwakftw** [comment]
@neerajprad do you think this is good to merge?

**Review by neerajprad — APPROVED** [review]
This looks great! Thanks for exposing this cool sampler.

**Comment by neerajprad** [comment]
The failures are unrelated, so I am happy merging this! 

@pytorchbot merge this please

**Pull request closed without merging**